upload.py

```diff
@@ -1,18 +1,54 @@
-import os
```

Comment:
The changes in upload.py don't seem to be related to setup; were these changes committed by mistake, or do they belong in a separate commit?

Reply:
Yes they were included in the commit by mistake.

Reply:
As discussed over chat, changes to this file have been moved to a separate commit, but within the same branch on this occasion.